Add semantic segmentation example script, no trainer

NielsRogge opened the pull request:
# What does this PR do?

This PR adds an example script regarding fine-tuning any model supported by the `AutoModelForSemanticSegmentation` API on a semantic segmentation dataset, including regularly pushing to the hub during training as well as WandB logging.

I switched to using Accelerate as I had a bug with the Trainer. 

To do:

- [ ] fix Tensorboard logs

Commit:
Add first draft

Commit:
Add first draft

NielsRogge committed:
Add requirements.txt

NielsRogge committed:
Add requirements.txt

Commit:
Add wandb logging

Commit:
Add wandb logging

Commit:
Merge branch 'add_semantic_script_no_trainer' of https://github.com/NielsRogge/transformers into add_semantic_script_no_trainer

NielsRogge committed:
Improve script

NielsRogge committed:
Improve script

NielsRogge committed:
Add validation

NielsRogge committed:
Add validation

NielsRogge committed:
Improve script

NielsRogge committed:
Improve script

NielsRogge committed:
Improve README

NielsRogge committed:
Improve README

ydshieh committed:
Fix TFTransfoXLLMHeadModel outputs (#16590)

Co-authored-by: ydshieh <ydshieh@users.noreply.github.com>

sgugger committed:
Allow the same config in the auto mapping

sgugger committed:
Revert "Allow the same config in the auto mapping"

This reverts commit b1a7dfe099b852340868f9aa7c75bb805ce57596.

LysandreJik committed:
Dev version

stas00 committed:
[modeling_utils] rearrange text (#16632)

gante committed:
TF generate refactor - Beam Search (#16374)

* refactor TF beam search

* refactored generate can now properly use attention masks

* add force bos/eos logit processors

anmolsjoshi committed:
Added Annotations for PyTorch models (#16619)

* Update modeling_mpnet.py

* Update modeling_ctrl.py

* formatting

* Formatting

* Formatting

* annotated FSMT

* Added annotations for LED

* Added Annotations for M2M

* Added annotations for nystromformer

* Added annotations for OpenAI

* Added annotations for RAG

* Removed unused imports

* fix isort errors

* Removed inputs_embeds docstring, corrected original

* flake8 fixes

* doc-builder fixes

sgugger committed:
Allow the same config in the auto mapping (#16631)

muellerzr committed:
Update no_trainer scripts with new Accelerate functionalities (#16617)

Adds logging and save/loading to the Accelerate scripts

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

NielsRogge committed:
Fix doc example (#16448)

* Fix doc

* Make fixup

Co-authored-by: Niels Rogge <nielsrogge@nielss-mbp.home>

muellerzr commented on examples/pytorch/semantic-segmentation/run_semantic_segmentation_no_trainer.py:
```diff
@@ -0,0 +1,609 @@
+        eps=args.adam_epsilon,
+    )
+
+    # Prepare everything with our `accelerator`.
+    model, optimizer, train_dataloader, eval_dataloader = accelerator.prepare(
+        model, optimizer, train_dataloader, eval_dataloader
+    )
+
+    # Scheduler and math around the number of training steps.
+    num_update_steps_per_epoch = math.ceil(len(train_dataloader) / args.gradient_accumulation_steps)
+    if args.max_train_steps is None:
+        args.max_train_steps = args.num_train_epochs * num_update_steps_per_epoch
+    else:
+        args.num_train_epochs = math.ceil(args.max_train_steps / num_update_steps_per_epoch)
+
+    lr_scheduler = get_scheduler(
+        name=args.lr_scheduler_type,
+        optimizer=optimizer,
+        num_warmup_steps=args.num_warmup_steps,
+        num_training_steps=args.max_train_steps,
+    )
```
The lr scheduler now gets prepared by `Accelerate` as well
```suggestion
    # Scheduler and math around the number of training steps.
    num_update_steps_per_epoch = math.ceil(len(train_dataloader) / args.gradient_accumulation_steps)
    if args.max_train_steps is None:
        args.max_train_steps = args.num_train_epochs * num_update_steps_per_epoch
    else:
        args.num_train_epochs = math.ceil(args.max_train_steps / num_update_steps_per_epoch)

    lr_scheduler = get_scheduler(
        name=args.lr_scheduler_type,
        optimizer=optimizer,
        num_warmup_steps=args.num_warmup_steps,
        num_training_steps=args.max_train_steps,
    )
    
    # Prepare everything with our `accelerator`.
    model, optimizer, train_dataloader, eval_dataloader, lr_scheduler = accelerator.prepare(
        model, optimizer, train_dataloader, eval_dataloader, lr_scheduler
    )
```

muellerzr commented on examples/pytorch/semantic-segmentation/run_semantic_segmentation_no_trainer.py:
```diff
@@ -0,0 +1,609 @@
+                    else:
+                        log_str += "| {}: {:.3e}".format(k, v)
+
+                if accelerator.is_local_main_process:
+                    progress_bar.write(log_str)
+                    if is_wandb_available() and args.wandb:
+                        wandb.log(train_logs)
```
This can be refactored some based on your preference. I recommend something like this:
```suggestion
                if accelerator.is_local_main_process:
                    progress_bar.write(log_str)
                if args.wandb:
                    accelerator.log(train_logs)
```

muellerzr reviewed (changes requested):
Thanks @NielsRogge!

I added a few suggestions based on the recent upgrades to Accelerate and the new script changes that got pushed. 

The other two capabilities we're trying to include in the scripts are [resuming from a checkpoint](https://github.com/huggingface/transformers/blob/main/examples/pytorch/multiple-choice/run_swag_no_trainer.py#L500) as well as `save_state` by [batch](https://github.com/huggingface/transformers/blob/main/examples/pytorch/multiple-choice/run_swag_no_trainer.py#L541) or by [epoch](https://github.com/huggingface/transformers/blob/main/examples/pytorch/question-answering/run_qa_no_trainer.py#L882) (this is just `torch.save`, so slightly different than transformers save. Pick your poison on those I think.)

Nice start!

lmvasque committed:
Add inputs vector to calculate metric method (#16461)

* Add inputs vector to calculate metric method

* Include inputs for evaluation metrics with backwards compatibility

* Prevent inputs create OOM issue and documentation details

* Update style and code documentation

* Fix style formatting issues

* Update files format with make style

sgugger commented on examples/pytorch/semantic-segmentation/run_semantic_segmentation_no_trainer.py:
```diff
@@ -0,0 +1,609 @@
+        transformers.utils.logging.set_verbosity_error()
+
+    # If passed along, set the training seed now.
+    if args.seed is not None:
+        set_seed(args.seed)
```
They were written at the time the one in Accelerate was the same. You should definitely use the one from accelerate with  `device_specific=True` since you have some data augmentation here (otherwise all GPUs will do the exact same data augmentation).

muellerzr replied:
Hmm... that might be something that needs adjusting then, as the Accelerate one is different 😄 

cc @sgugger

sgugger commented on examples/pytorch/semantic-segmentation/run_semantic_segmentation_no_trainer.py:
```diff
@@ -0,0 +1,609 @@
+        args.model_name_or_path, num_labels=num_labels, id2label=id2label, label2id=label2id
+    )
+    feature_extractor = AutoFeatureExtractor.from_pretrained(args.model_name_or_path)
+    model = AutoModelForSemanticSegmentation.from_pretrained(
+        args.model_name_or_path,
+        config=config,
+    )
```
```suggestion
    model = AutoModelForSemanticSegmentation.from_pretrained(args.model_name_or_path, config=config)
```

sgugger reviewed (approved):
Very nice addition! Agreed with all the changes Zach suggested as Accelerate can now handle the logging for you, and the scheduler should be passed along to `prepare` for better support with DeepSpeed.

This will make your example rely on Accelerate master for a couple of weeks until next release (probably next week) so you should add a warning in the Readme to install Accelerate from source. I'll put a not to adapt this after Accelerate next release.

stas00 committed:
[megatron-bert-uncased-345m] fix conversion (#16639)

sgugger committed:
Remove parent/child tests in auto model tests (#16653)

NielsRogge committed:
Add Accelerate new features

NielsRogge committed:
Add Accelerate new features

FrancescoSaverioZuppichini committed:
Updated _load_pretrained_model_low_mem to check if keys are in the state_dict (#16643)

* Updated _load_pretrained_model_low_mem to check if keys are in the stored state_dict

* update after conversions

BritneyMuller committed:
Update Support image on README.md (#16615)

* Update README.md Support Image

Updates the Support image linking to our EAP page (to give it a refresh + help avoid image fatigue).

Slack thread checking in with #open-source-internal on this update (https://huggingface.slack.com/archives/C021H1P1HKR/p1648838903316709)

* Compressed Updated Support image

* Improves Support Image Logo + Height

Updated the image based on logo + size feedback. Big thanks to Bibi for making quick edits to this image.

FrancescoSaverioZuppichini committed:
RegNet (#16188)

* base model done

* make style

* done

* added files

* Apply suggestions from code review

Co-authored-by: NielsRogge <48327001+NielsRogge@users.noreply.github.com>

* Apply suggestions from code review

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

* Trigger doc build

* resolved conversations

* resolved conversations

* seer models

* minor changes

* minor changes

* make fixup

* glob variables

* minor changes

* fix copies

* config when possibile

* resolved conflicts

* resolved conflicts

* resolved conflicts

* CI

* conversion script for 10b param

* fixed for 10b model

* minor updates in the doc + make style

* removed unused code

* Apply suggestions from code review

Co-authored-by: NielsRogge <48327001+NielsRogge@users.noreply.github.com>

* removed unused code

* removed unused code

* updated modeling_utils from main

Co-authored-by: NielsRogge <48327001+NielsRogge@users.noreply.github.com>
Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>
Co-authored-by: Sylvain Gugger <Sylvain.gugger@gmail.com>

stefan-it committed:
bert: properly mention deprecation of TF2 conversion script (#16171)

NielsRogge committed:
Add TAPEX (#16473)

* Add TapexTokenizer

* Improve docstrings and provide option to provide answer

* Remove option for pretokenized inputs

* Add TAPEX to README

* Fix copies

* Remove option for pretokenized inputs

* Initial commit: add tapex fine-tuning examples on both table-based question answering and table-based fact verification.

* - Draft a README file for running the script and introducing some background.
- Remove unused code lines in tabfact script.
- Disable the deafult `pad_to_max_length` option which is memory-consuming.

* * Support `as_target_tokenizer` function for TapexTokenizer.
* Fix the do_lower_case behaviour of TapexTokenizer.
* Add unit tests for target scenarios and cased/uncased scenarios for both source and target.

* * Replace the label BartTokenizer with TapexTokenizer's as_target_tokenizer function.
* Fix typos in tapex example README.

* * fix the evaluation script - remove the property `task_name`

* * Make the label space more clear for tabfact tasks

* * Using a new fine-tuning script for tapex-base on tabfact.

* * Remove the lowercase code outside the tokenizer - we use the tokenizer to control whether do_lower_case
* Guarantee the hyper-parameter can be run without out-of-memory on 16GB card and report the new reproduced number on wikisql

* * Remove the default tokenizer_name option.
* Provide evaluation command.

* * Support for WikiTableQuestion dataset.

* Fix a typo in README.

* * Fix the datasets's key name in WikiTableQuestions

* Run make fixup and move test to folder

* Fix quality

* Apply suggestions from code review

* Apply suggestions from code review

Co-authored-by: Suraj Patil <surajp815@gmail.com>

* Apply suggestions from code review

* Apply suggestions from code review

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

* Apply some more suggestions from code review

* Improve docstrings

* Overwrite failing test

* Improve comment in example scripts

* Fix rebase

* Add TAPEX to Auto mapping

* Add TAPEX to auto config mappings

* Put TAPEX higher than BART in auto mapping

* Add TAPEX to doc tests

Co-authored-by: Niels Rogge <nielsrogge@Nielss-MBP.localdomain>
Co-authored-by: SivilTaram <qianlxc@outlook.com>
Co-authored-by: Niels Rogge <nielsrogge@nielss-mbp.home>
Co-authored-by: Suraj Patil <surajp815@gmail.com>
Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>
Co-authored-by: Niels Rogge <nielsrogge@Nielss-MacBook-Pro.local>

johko committed:
add vit tf doctest with @add_code_sample_docstrings (#16636)

* add vit tf doctest with @add_code_sample_docstrings

* add labels string back in

Co-authored-by: Johannes Kolbe <johannes.kolbe@tech.better.team>

secsilm committed:
Fix error in doc of `DataCollatorWithPadding` (#16662)

The defalut value of `padding` in `DataCollatorWithPadding` is `True`, not `False`.

sgugger committed:
Fix style

ydshieh committed:
Fix QA sample (#16648)

* fix QA sample

* For TF_QUESTION_ANSWERING_SAMPLE

Co-authored-by: ydshieh <ydshieh@users.noreply.github.com>

muellerzr committed:
Add tests for no_trainer and fix existing examples (#16656)

* Fixed some bugs involving saving during epochs
* Added tests mimicking the existing examples tests
* Added in json exporting to all `no_trainer` examples for consistency

NielsRogge committed:
Improve script

NielsRogge committed:
Improve script

laurahanu committed:
only load state dict when the checkpoint is not None (#16673)

stas00 committed:
[Trainer] tf32 arg doc (#16674)

* [Trainer] tf32 arg doc

* Update src/transformers/training_args.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

stevhliu committed:
Update audio examples with MInDS-14 (#16633)

* ✨ update audio examples with minds dataset

* 🖍 make style

* 🖍 minor fixes for doctests

NielsRogge commented:
I've added all requested changes. As you can see, I'm updating the repo using commits rather than creating subfolders.

sgugger commented on examples/pytorch/semantic-segmentation/run_semantic_segmentation_no_trainer.py:
```diff
@@ -0,0 +1,637 @@
+                    # Optionally push to the hub
+                    if args.push_to_hub and accelerator.is_main_process:
+                        feature_extractor.save_pretrained(args.output_dir)
+                        repo.push_to_hub(
```
Note that the checkpoint saved here is saved with `torch.save` and not `save_pretrained` so they're not usable with `from_pretrained`.

Also, you'll need to adapt the logic to resume from a checkpoint to checkout the right branch in the folder if you go this way, as everything is saved in the root of `output_dir`.

NielsRogge replied:
I've removed this logic and updated the script to be aligned with the other example scripts.

NielsRogge committed:
Add corresponding test

NielsRogge committed:
Fix merge

NielsRogge committed:
Make script more aligned to other scripts

NielsRogge commented:
Some issue with git, closing this issue in favor of a new one.
